Cython/Compiler/Code.py

```diff
@@ -1216,10 +1216,12 @@ def initialize_main_c_code(self):
         code.putln("/* --- Runtime support code --- */")
 
     def initialize_main_h_code(self):
-        # h code can include utility code in a few cases so it needs to be insertable
+        # h files can include only include a much smaller list of sections
+        # that are manually created here
         rootwriter = self.rootwriter
         for part in self.code_layout:
-            if part.find("utility_code") != -1:
+            if part in ('h_code', 'utility_code_proto_before_types', 'type_declarations',
+                        'utility_code_proto', 'end'):
                 self.parts[part] = rootwriter.insertion_point()
```

Comment:
If we know all part names, why not create them explicitly? Why do we need the loop over `code_layout` here?

Suggestion: Create `c_code_layout` and `h_code_layout` in `Code.py`, next to each other, so that they are easy to keep in sync. Then having the two initialisation functions seems reasonable. My line of reasoning is that the header file generation is not a "special case" or a strict subset of the C file generation. They are at the same level. The implementation should reflect that.